pymc3/distributions/discrete.py

```diff
@@ -700,22 +706,23 @@ def __init__(self, p, *args, **kwargs):
             self.k = tt.shape(p)[-1].tag.test_value
         except AttributeError:
             self.k = tt.shape(p)[-1]
-        self.p = p = tt.as_tensor_variable(p)
+        p = tt.as_tensor_variable(p)
         self.p = (p.T / tt.sum(p, -1)).T
         self.mode = tt.argmax(p)
 
-    def random(self, point=None, size=None):
-        def random_choice(k, *args, **kwargs):
-            if len(kwargs['p'].shape) > 1:
-                return np.asarray(
-                    [np.random.choice(k, p=p)
-                     for p in kwargs['p']]
-                )
-            else:
-                return np.random.choice(k, *args, **kwargs)
+    def _random(self, k, p, size=None):
+        if len(p.shape) > 1:
+            return np.asarray(
+                [np.random.choice(k, p=p, size=size)
+                    for p in kwargs['p']]
```

Comment:
should this be `p` instead of `kwargs['p']`?

Reply:
good catch! seems to indicate this path was never hit though...

Reply:
You should change it otherwise below fail:
```python
p = np.random.rand(10, 5)
pm.Categorical.dist(p=p).random()
```

 Might be worth to add a test case.

Reply:
Ack, I promise I changed it and just forgot to save/add to commit!  Just did, though.

Reply:
Actually, the random in Categorical still returns wrong shape, running:
```
p = np.random.rand(10, 5)
pm.Categorical.dist(p=p).random()
```
gives a (10, 5) instead of (10, ).

Reply:
but i think we can clean that up in a separate PR as well.

Reply:
`p` has shape (10, 5), though, so I think that makes sense?

Reply:
ooh, yeah, but at least documenting that the last dimension of `p` is used for the number of categories, and maybe more of the broadcasting rules might be nice.